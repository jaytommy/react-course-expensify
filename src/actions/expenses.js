import database from '../firebase/firebase';

/* Originally we had the action generator return an object.  Now using Firebase,
we change to have it return a function.  Then the component will dispacth the function.
The function will then run with the ability to dispatch other actions including the
Firebase write calls.
*/

// ADD_EXPENSE
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = '', 
      note = '', 
      amount = 0, 
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt };
    
    return database.ref('expenses').push(expense)
      .then((ref) => {
        dispatch(addExpense({
          id: ref.key,
          ...expense
        }));
      });
  };
};

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
type: 'REMOVE_EXPENSE',
id
});

export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).remove().then(() => {
      dispatch(removeExpense({ id }));
    });
  };
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
type: 'EDIT_EXPENSE',
id,
updates
});

export const startEditExpense = (id, updates) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).update(updates).then(() => {
      dispatch(editExpense(id, updates));
    });
  };
};

// SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

export const startSetExpenses = () => {
  // need to return a function
  return (dispatch) => {
    return database.ref('expenses').once('value').then((snapshot) => {
      const expenses = [];
      snapshot.forEach((childSnapshot) => {
        expenses.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      });
      dispatch(setExpenses(expenses));
    });
  };
};