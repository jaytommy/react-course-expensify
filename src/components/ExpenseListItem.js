// Render description, amount, createdAt
import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral'

// Destructure props to three components 
const ExpenseListItem = ({ id, description, amount, createdAt }) => (
  <div>
      <Link to={`/edit/${id}`}>
        <h3>{description}</h3>
      </Link>
      <p>
        Amount: {numeral(amount / 100).format('$0,0.00')}
        
        Created: {moment(createdAt).format('MMMM Do, YYYY')}
      </p>
  </div>
);

export default ExpenseListItem;