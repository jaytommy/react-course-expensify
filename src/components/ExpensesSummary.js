import React from 'react';
import numeral from 'numeral';
import { connect } from 'react-redux';
import selectExpensesTotal from '../selectors/expenses-total';
import selectExpenses from '../selectors/expenses';

// Connect to store for:
// -- expenseCount (visible expenses)
// -- expensesTotal

export const ExpensesSummary = ({ expenseCount, expensesTotal}) => {
  const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
  return (
    <div>
      <h1>Viewing {expenseCount} {expenseWord} totalling {expensesTotal}</h1>
    </div>
  );
};


const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  return {
    expenseCount: visibleExpenses.length,
    expensesTotal: numeral(selectExpensesTotal(visibleExpenses) / 100).format('$0,0.00')
  };
};

export default connect(mapStateToProps)(ExpensesSummary);