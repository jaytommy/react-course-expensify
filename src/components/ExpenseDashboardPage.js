import React from 'react';
import ExpensesSummary from './ExpensesSummary.js';
import ExpenseList from './ExpenseList.js';
import ExpenseListFilters from './ExpenseListFilters.js';

const ExpenseDashboardPage = () => (
  <div>
    <ExpensesSummary />
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;