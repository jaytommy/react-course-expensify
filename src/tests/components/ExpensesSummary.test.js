import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

// Snapshot test 1
test('should render ExpenseSummary with one expense', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={'$235.75'}/>);
  expect(wrapper).toMatchSnapshot();
});

// // Snapshot test 2
test('should render ExpenseSummary with expenses and filters', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={23} expensesTotal={'$1346235.75'}/>);
  expect(wrapper).toMatchSnapshot();
});