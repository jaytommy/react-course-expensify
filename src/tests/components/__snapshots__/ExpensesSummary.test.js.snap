// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseSummary with expenses and filters 1`] = `
<div>
  <h1>
    Viewing 
    23
     
    expenses
     totalling 
    $1346235.75
  </h1>
</div>
`;

exports[`should render ExpenseSummary with one expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totalling 
    $235.75
  </h1>
</div>
`;
